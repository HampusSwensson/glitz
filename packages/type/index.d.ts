import * as CSS from 'csstype';

export interface Style extends FeaturedProperties, PseudoMap {}

export interface Properties extends CSS.StandardLonghandPropertiesFallback, CSS.VendorPropertiesFallback {}

export interface FeaturedProperties
  extends Omit<Properties, keyof ExtendedProperties>,
    ExtendedProperties,
    Omit<CSS.StandardShorthandPropertiesFallback, keyof ShorthandProperties>,
    ShorthandProperties {
  '@keyframes'?: PropertiesList;
  '@font-face'?: FontFace;
}

export interface ExtendedProperties {
  animationName?: PropertiesList | Properties['animationName'];
  fontFamily?: FontFamily;
}

export interface ShorthandProperties {
  animation?: Animation;
  background?: Background;
  border?: Border;
  borderBottom?: BorderBottom;
  borderImage?: BorderImage;
  borderLeft?: BorderLeft;
  borderRight?: BorderRight;
  borderTop?: BorderTop;
  flex?: Flex;
  font?: Font;
  grid?: Grid;
  maskBorder?: MaskBorder;
  mask?: Mask;
  margin?: Margin;
  offset?: Offset;
  outline?: Outline;
  padding?: Padding;
  transition?: Transition;
}

interface Animation {
  delay?: CSS.StandardLonghandPropertiesFallback['animationDelay'];
  direction?: CSS.StandardLonghandPropertiesFallback['animationDirection'];
  duration?: CSS.StandardLonghandPropertiesFallback['animationDuration'];
  fillMode?: CSS.StandardLonghandPropertiesFallback['animationFillMode'];
  iterationCount?: CSS.StandardLonghandPropertiesFallback['animationIterationCount'];
  name?: PropertiesList | CSS.StandardLonghandPropertiesFallback['animationName'];
  playState?: CSS.StandardLonghandPropertiesFallback['animationPlayState'];
  timingFunction?: CSS.StandardLonghandPropertiesFallback['animationTimingFunction'];
}

interface Background {
  attachment?: CSS.StandardLonghandPropertiesFallback['backgroundAttachment'];
  blendMode?: CSS.StandardLonghandPropertiesFallback['backgroundBlendMode'];
  clip?: CSS.StandardLonghandPropertiesFallback['backgroundClip'];
  color?: CSS.StandardLonghandPropertiesFallback['backgroundColor'];
  image?: CSS.StandardLonghandPropertiesFallback['backgroundImage'];
  origin?: CSS.StandardLonghandPropertiesFallback['backgroundOrigin'];
  position?: CSS.StandardLonghandPropertiesFallback['backgroundPosition'];
  repeat?: CSS.StandardLonghandPropertiesFallback['backgroundRepeat'];
  size?: CSS.StandardLonghandPropertiesFallback['backgroundSize'];
}

interface Border {
  collapse?: CSS.StandardLonghandPropertiesFallback['borderCollapse'];
  color?: CSS.StandardShorthandPropertiesFallback['borderColor'];
  blockEnd?: CSS.StandardShorthandPropertiesFallback['borderBlockEnd'];
  blockStart?: CSS.StandardShorthandPropertiesFallback['borderBlockStart'];
  inlineEnd?: CSS.StandardShorthandPropertiesFallback['borderInlineEnd'];
  inlineStart?: CSS.StandardShorthandPropertiesFallback['borderInlineStart'];
  radius?: CSS.StandardShorthandPropertiesFallback['borderRadius'];
  spacing?: CSS.StandardLonghandPropertiesFallback['borderSpacing'];
  style?: CSS.StandardShorthandPropertiesFallback['borderStyle'];
  width?: CSS.StandardShorthandPropertiesFallback['borderWidth'];
}

interface BorderBottom {
  color?: CSS.StandardLonghandPropertiesFallback['borderBottomColor'];
  style?: CSS.StandardLonghandPropertiesFallback['borderBottomStyle'];
  width?: CSS.StandardLonghandPropertiesFallback['borderBottomWidth'];
}

interface BorderImage {
  outset?: CSS.StandardLonghandPropertiesFallback['borderImageOutset'];
  repeat?: CSS.StandardLonghandPropertiesFallback['borderImageRepeat'];
  slice?: CSS.StandardLonghandPropertiesFallback['borderImageSlice'];
  source?: CSS.StandardLonghandPropertiesFallback['borderImageSource'];
  width?: CSS.StandardLonghandPropertiesFallback['borderImageWidth'];
}

interface BorderLeft {
  color?: CSS.StandardLonghandPropertiesFallback['borderLeftColor'];
  style?: CSS.StandardLonghandPropertiesFallback['borderLeftStyle'];
  width?: CSS.StandardLonghandPropertiesFallback['borderLeftWidth'];
}

interface BorderRight {
  color?: CSS.StandardLonghandPropertiesFallback['borderRightColor'];
  style?: CSS.StandardLonghandPropertiesFallback['borderRightStyle'];
  width?: CSS.StandardLonghandPropertiesFallback['borderRightWidth'];
}

interface BorderTop {
  color?: CSS.StandardLonghandPropertiesFallback['borderTopColor'];
  style?: CSS.StandardLonghandPropertiesFallback['borderTopStyle'];
  width?: CSS.StandardLonghandPropertiesFallback['borderTopWidth'];
}

interface Flex {
  basis?: CSS.StandardLonghandPropertiesFallback['flexBasis'];
  direction?: CSS.StandardLonghandPropertiesFallback['flexDirection'];
  grow?: CSS.StandardLonghandPropertiesFallback['flexGrow'];
  shrink?: CSS.StandardLonghandPropertiesFallback['flexShrink'];
  wrap?: CSS.StandardLonghandPropertiesFallback['flexWrap'];
}

interface Font {
  family?: FontFamily;
  featureSettings?: CSS.StandardLonghandPropertiesFallback['fontFeatureSettings'];
  kerning?: CSS.StandardLonghandPropertiesFallback['fontKerning'];
  languageOverride?: CSS.StandardLonghandPropertiesFallback['fontLanguageOverride'];
  variationSettings?: CSS.StandardLonghandPropertiesFallback['fontVariationSettings'];
  size?: CSS.StandardLonghandPropertiesFallback['fontSize'];
  sizeAdjust?: CSS.StandardLonghandPropertiesFallback['fontSizeAdjust'];
  stretch?: CSS.StandardLonghandPropertiesFallback['fontStretch'];
  style?: CSS.StandardLonghandPropertiesFallback['fontStyle'];
  synthesis?: CSS.StandardLonghandPropertiesFallback['fontSynthesis'];
  variant?: CSS.StandardLonghandPropertiesFallback['fontVariant'];
  weight?: CSS.StandardLonghandPropertiesFallback['fontWeight'];
}

interface Grid {
  autoColumns?: CSS.StandardLonghandPropertiesFallback['gridAutoColumns'];
  autoFlow?: CSS.StandardLonghandPropertiesFallback['gridAutoFlow'];
  autoRows?: CSS.StandardLonghandPropertiesFallback['gridAutoRows'];
  column?: CSS.StandardShorthandPropertiesFallback['gridColumn'];
  columnGap?: CSS.StandardLonghandPropertiesFallback['gridColumnGap'];
  row?: CSS.StandardShorthandPropertiesFallback['gridRow'];
  rowGap?: CSS.StandardLonghandPropertiesFallback['gridRowGap'];
  template?: CSS.StandardShorthandPropertiesFallback['gridTemplate'];
}

interface MaskBorder {
  mode?: CSS.StandardLonghandPropertiesFallback['maskBorderMode'];
  outset?: CSS.StandardLonghandPropertiesFallback['maskBorderOutset'];
  repeat?: CSS.StandardLonghandPropertiesFallback['maskBorderRepeat'];
  slice?: CSS.StandardLonghandPropertiesFallback['maskBorderSlice'];
  source?: CSS.StandardLonghandPropertiesFallback['maskBorderSource'];
  width?: CSS.StandardLonghandPropertiesFallback['maskBorderWidth'];
}

interface Mask {
  clip?: CSS.StandardLonghandPropertiesFallback['maskClip'];
  composite?: CSS.StandardLonghandPropertiesFallback['maskComposite'];
  image?: CSS.StandardLonghandPropertiesFallback['maskImage'];
  mode?: CSS.StandardLonghandPropertiesFallback['maskMode'];
  origin?: CSS.StandardLonghandPropertiesFallback['maskOrigin'];
  position?: CSS.StandardLonghandPropertiesFallback['maskPosition'];
  repeat?: CSS.StandardLonghandPropertiesFallback['maskRepeat'];
  size?: CSS.StandardLonghandPropertiesFallback['maskSize'];
  type?: CSS.StandardLonghandPropertiesFallback['maskType'];
}

interface Margin {
  x?: CSS.StandardLonghandPropertiesFallback['marginLeft'];
  y?: CSS.StandardLonghandPropertiesFallback['marginTop'];
  top?: CSS.StandardLonghandPropertiesFallback['marginTop'];
  right?: CSS.StandardLonghandPropertiesFallback['marginRight'];
  bottom?: CSS.StandardLonghandPropertiesFallback['marginBottom'];
  left?: CSS.StandardLonghandPropertiesFallback['marginLeft'];
}

interface Offset {
  anchor?: CSS.StandardLonghandPropertiesFallback['offsetAnchor'];
  blockEnd?: CSS.StandardLonghandPropertiesFallback['offsetBlockEnd'];
  blockStart?: CSS.StandardLonghandPropertiesFallback['offsetBlockStart'];
  inlineEnd?: CSS.StandardLonghandPropertiesFallback['offsetInlineEnd'];
  inlineStart?: CSS.StandardLonghandPropertiesFallback['offsetInlineStart'];
  distance?: CSS.StandardLonghandPropertiesFallback['offsetDistance'];
  path?: CSS.StandardLonghandPropertiesFallback['offsetPath'];
  position?: CSS.StandardLonghandPropertiesFallback['offsetPosition'];
  rotate?: CSS.StandardLonghandPropertiesFallback['offsetRotate'];
}

interface Outline {
  color?: CSS.StandardLonghandPropertiesFallback['outlineColor'];
  offset?: CSS.StandardLonghandPropertiesFallback['outlineOffset'];
  style?: CSS.StandardLonghandPropertiesFallback['outlineStyle'];
  width?: CSS.StandardLonghandPropertiesFallback['outlineWidth'];
}

interface Padding {
  x?: CSS.StandardLonghandPropertiesFallback['paddingLeft'];
  y?: CSS.StandardLonghandPropertiesFallback['paddingTop'];
  top?: CSS.StandardLonghandPropertiesFallback['paddingTop'];
  right?: CSS.StandardLonghandPropertiesFallback['paddingRight'];
  bottom?: CSS.StandardLonghandPropertiesFallback['paddingBottom'];
  left?: CSS.StandardLonghandPropertiesFallback['paddingLeft'];
}

interface Transition {
  delay?: CSS.StandardLonghandPropertiesFallback['transitionDelay'];
  duration?: CSS.StandardLonghandPropertiesFallback['transitionDuration'];
  property?: CSS.StandardLonghandPropertiesFallback['transitionProperty'];
  timingFunction?: CSS.StandardLonghandPropertiesFallback['transitionTimingFunction'];
}

export type PseudoMap = { [P in CSS.SimplePseudos]?: FeaturedProperties & PseudoMap };

export interface PropertiesList {
  [identifier: string]: Properties;
}

export type FontFamily =
  | FontFace
  | CSS.StandardLonghandProperties['fontFamily']
  | Array<FontFace | CSS.StandardLonghandProperties['fontFamily']>;

export type FontFace = Omit<CSS.FontFaceFallback, 'fontFamily'>;

type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];
type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;
