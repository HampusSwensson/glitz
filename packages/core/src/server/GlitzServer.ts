import { Style } from '@glitz/type';
import Base from '../core/Base';
import { DEFAULT_HYDRATION_IDENTIFIER, Options } from '../types/options';
import { createHashCounter } from '../utils/hash';
import InjectorServer from './InjectorServer';

export default class GlitzServer<TStyle = Style> extends Base<TStyle> {
  public getStyleMarkup: (className?: string) => string;
  constructor(options: Options = {}) {
    const prefix = options.prefix;
    const classHasher = createHashCounter(prefix);
    const keyframesHasher = createHashCounter(prefix);
    const fontFaceHasher = createHashCounter(prefix);

    let plain: InjectorServer;
    const mediaIndex: {
      [media: string]: InjectorServer;
    } = {};

    const injector = (media?: string) =>
      media
        ? (mediaIndex[media] = mediaIndex[media] || new InjectorServer(classHasher, keyframesHasher, fontFaceHasher))
        : (plain = plain || new InjectorServer(classHasher, keyframesHasher, fontFaceHasher));

    super(injector, options.transformer, options.atomic);

    this.getStyleMarkup = () => {
      const attribute = options.identifier || DEFAULT_HYDRATION_IDENTIFIER;
      let markup = '';
      if (plain) {
        markup += `<style data-${attribute}>${plain.getStyle()}</style>`;
      }
      const medias = options.mediaOrder ? Object.keys(mediaIndex).sort(options.mediaOrder) : Object.keys(mediaIndex);
      for (const media of medias) {
        markup += `<style data-${attribute} media="${media}">${mediaIndex[media].getStyle()}</style>`;
      }
      return markup;
    };
  }
}
