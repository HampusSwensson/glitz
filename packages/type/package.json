{
  "name": "@glitz/type",
  "version": "1.1.1",
  "main": "",
  "types": "./index.d.ts",
  "license": "MIT",
  "author": "Fredrik Nicol <fredrik.nicol@gmail.com>",
  "homepage": "https://github.com/frenic/glitz",
  "repository": "git@github.com:frenic/glitz.git",
  "bugs": "https://github.com/frenic/glitz/issues",
  "dependencies": {
    "csstype": "^1.8.1"
  },
  "files": [
    "index.d.ts"
  ]
}
