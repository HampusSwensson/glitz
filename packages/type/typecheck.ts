// tslint:disable

import { Style } from './';

const css: Style = {
  gridColumnStart: '',
  grid: {
    column: {
      start: '',
    },
  },
  ':hover': {
    gridColumnStart: '',
    grid: {
      column: {
        start: '',
      },
    },
  },
  '@keyframes': {
    from: {
      gridColumnStart: '',
      grid: {
        column: {
          start: '',
        },
      },
    },
  },
  animationName: {
    from: {
      gridColumnStart: '',
      grid: {
        column: {
          start: '',
        },
      },
    },
  },
  '@font-face': {
    fontVariant: [''],
    font: {
      variant: '',
    },
  },
  fontFamily: {
    fontVariant: [''],
    font: {
      variant: '',
    },
  },
};

// Avoid unread variables type error
css;
