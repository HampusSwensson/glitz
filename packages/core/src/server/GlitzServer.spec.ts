import { Properties, Style } from '@glitz/type';
import GlitzServer from './GlitzServer';

interface TestStyle extends Style {
  '@media (min-width: 100px)'?: Style;
  '@media (min-width: 200px)'?: Style;
  '@media (min-width: 300px)'?: Style;
  '@media (min-width: 768px)'?: Style;
  '@media (min-width: 992px)'?: Style;
}

describe('server', () => {
  it('injects plain rule', () => {
    const server = new GlitzServer<TestStyle>();

    expect(server.injectStyle({ color: 'red' })).toBe('a');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects shorthand rule', () => {
    const server = new GlitzServer<TestStyle>();

    expect(
      server.injectStyle({
        padding: { left: '10px', right: '10px', top: '10px', bottom: '10px' },
      }),
    ).toBe('a b c d');

    expect(
      server.injectStyle({
        grid: { column: { gap: '10px' } },
      }),
    ).toBe('e');

    expect(
      server.injectStyle({
        margin: { x: '10px' },
      }),
    ).toBe('f g');

    expect(
      server.injectStyle({
        margin: { y: '10px' },
      }),
    ).toBe('h i');

    expect(
      server.injectStyle({
        margin: { xy: '20px' },
      }),
    ).toBe('j k l m');

    expect(
      server.injectStyle({
        padding: { left: '20px' },
        paddingLeft: '30px',
      }),
    ).toBe('n');

    expect(
      server.injectStyle({
        animationName: { from: { padding: { left: '20px' } } },
      }),
    ).toBe('o');

    expect(
      server.injectStyle({
        fontFamily: { font: { weight: 'bold' } },
      }),
    ).toBe('p');

    expect(
      server.injectStyle({
        border: { xy: { width: 0 }, x: { color: 'red' }, y: { color: 'green' } },
      }),
    ).toBe('q r s t u v w x');

    expect(
      server.injectStyle({
        border: { y: { x: { radius: 0 } } },
      }),
    ).toBe('y z a0 a1');

    expect(
      server.injectStyle({
        border: { x: { width: '10px' }, left: { width: '20px' } },
      }),
    ).toBe('a2 a3');

    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects pseudo rule', () => {
    const server = new GlitzServer<TestStyle>();

    expect(server.injectStyle({ color: 'red' })).toBe('a');
    expect(server.injectStyle({ ':hover': { color: 'red' } })).toBe('b');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects nested pseudo rule', () => {
    const server = new GlitzServer<TestStyle>();

    expect(server.injectStyle({ ':first-child': { ':hover': { color: 'red' } } })).toBe('a');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects media rule', () => {
    const server = new GlitzServer<TestStyle>();

    expect(server.injectStyle({ color: 'red' })).toBe('a');
    expect(server.injectStyle({ ':hover': { color: 'red' } })).toBe('b');
    expect(server.injectStyle({ '@media (min-width: 768px)': { color: 'red' } })).toBe('c');
    expect(server.injectStyle({ '@media (min-width: 768px)': { ':hover': { color: 'red' } } })).toBe('d');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects media markup in certain order', () => {
    const server = new GlitzServer<TestStyle>({
      mediaOrder: (a: string, b: string) => {
        if (a < b) {
          return -1;
        }
        if (a > b) {
          return 1;
        }
        return 0;
      },
    });

    expect(server.injectStyle({ '@media (min-width: 300px)': { color: 'red' } })).toBe('a');
    expect(server.injectStyle({ '@media (min-width: 100px)': { color: 'red' } })).toBe('b');
    expect(server.injectStyle({ '@media (min-width: 200px)': { color: 'red' } })).toBe('c');
    expect(server.injectStyle({ color: 'red' })).toBe('d');

    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects atomic rules', () => {
    const server = new GlitzServer<TestStyle>();

    expect(
      server.injectStyle({
        color: 'red',
        background: { color: 'green' },
        borderLeftColor: 'blue',
        ':hover': { color: 'red', background: { color: 'green' }, borderLeftColor: 'blue' },
        '@media (min-width: 768px)': { color: 'red', background: { color: 'green' }, borderLeftColor: 'blue' },
      }),
    ).toBe('a b c d e f g h i');

    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects non-atomic rules', () => {
    const server = new GlitzServer<TestStyle>({ atomic: false });

    expect(
      server.injectStyle({
        color: 'red',
        background: { color: 'green' },
        borderLeftColor: 'blue',
        ':hover': { color: 'red', background: { color: 'green' }, borderLeftColor: 'blue' },
        '@media (min-width: 768px)': { color: 'red', background: { color: 'green' }, borderLeftColor: 'blue' },
      }),
    ).toBe('a b c');

    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects keyframes rule', () => {
    const server = new GlitzServer<TestStyle>();

    expect(server.injectStyle({ '@keyframes': { from: { color: 'red' }, to: { color: 'green' } } })).toBe('a');
    expect(server.injectStyle({ animationName: { from: { color: 'blue' }, to: { color: 'white' } } })).toBe('b');
    expect(server.injectStyle({ animation: { name: { from: { color: 'blue' }, to: { color: 'white' } } } })).toBe('b');
    expect(server.injectStyle({ animationName: 'some-thing' })).toBe('c');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects font face rule', () => {
    const server = new GlitzServer<TestStyle>();

    expect(
      server.injectStyle({
        '@font-face': {
          fontStyle: 'normal',
          fontWeight: 400,
          src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2')",
        },
      }),
    ).toBe('a');

    expect(
      server.injectStyle({
        fontFamily: {
          fontStyle: 'normal',
          fontWeight: 400,
          src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
        },
      }),
    ).toBe('b');

    expect(
      server.injectStyle({
        font: {
          family: {
            fontStyle: 'normal',
            fontWeight: 400,
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
          },
        },
      }),
    ).toBe('b');

    expect(
      server.injectStyle({
        fontFamily: [
          {
            fontStyle: 'normal',
            fontWeight: 400,
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
          },
          'sans-serif',
        ],
      }),
    ).toBe('c');

    expect(
      server.injectStyle({
        font: {
          family: [
            {
              fontStyle: 'normal',
              fontWeight: 400,
              src:
                "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
            },
            'sans-serif',
          ],
        },
      }),
    ).toBe('c');

    expect(
      server.injectStyle({
        fontFamily: 'sans-serif',
      }),
    ).toBe('d');

    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects different combinations', () => {
    const server = new GlitzServer<TestStyle>();

    expect(
      server.injectStyle({
        color: 'red',
        '@media (min-width: 768px)': { color: 'green' },
        '@media (min-width: 992px)': { color: 'blue' },
      }),
    ).toBe('a b c');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('injects rule deeply', () => {
    const server = new GlitzServer<TestStyle>();

    expect(server.injectStyle([{ color: 'green' }, { color: 'red' }])).toBe('a');

    expect(
      server.injectStyle([
        {
          padding: { left: '10px' },
        },
        {
          paddingLeft: '20px',
        },
      ]),
    ).toBe('b');

    expect(server.injectStyle([{ ':hover': { color: 'green' } }, { ':hover': { color: 'red' } }])).toBe('c');
    expect(
      server.injectStyle([
        { ':first-child': { ':hover': { color: 'green' } } },
        { ':first-child': { ':hover': { color: 'red' } } },
      ]),
    ).toBe('d');

    expect(
      server.injectStyle([
        { '@keyframes': { from: { color: 'red' }, to: { color: 'green' } } },
        { '@keyframes': { from: { color: 'green' }, to: { color: 'blue' } } },
      ]),
    ).toBe('e');
    expect(
      server.injectStyle([
        { animationName: { from: { color: 'blue' }, to: { color: 'white' } } },
        { animationName: { from: { color: 'white' }, to: { color: 'black' } } },
      ]),
    ).toBe('f');

    expect(
      server.injectStyle([
        {
          '@font-face': {
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2')",
          },
        },
        {
          '@font-face': {
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
          },
        },
      ]),
    ).toBe('g');
    expect(
      server.injectStyle([
        {
          '@font-face': {
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2')",
          },
        },
        {
          '@font-face': {
            src: "url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2')",
          },
        },
      ]),
    ).toBe('h');

    expect(
      server.injectStyle([
        { '@media (min-width: 768px)': { color: 'green' } },
        { '@media (min-width: 768px)': { color: 'red' } },
      ]),
    ).toBe('i');
    expect(
      server.injectStyle([
        { '@media (min-width: 768px)': { ':hover': { color: 'green' } } },
        { '@media (min-width: 768px)': { ':hover': { color: 'red' } } },
      ]),
    ).toBe('j');

    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('preserves order', () => {
    const atomic = new GlitzServer<TestStyle>();

    expect(
      atomic.injectStyle({
        padding: { left: '20px', right: '20px', top: '20px' },
        color: 'red',
        paddingRight: '30px',
        ':hover': {
          color: 'green',
        },
        paddingLeft: '30px',
      }),
    ).toBe('a b c d e');

    expect(atomic.getStyleMarkup()).toMatchSnapshot();

    const nonAtomic = new GlitzServer<TestStyle>({ atomic: false });

    expect(
      nonAtomic.injectStyle({
        padding: { left: '20px', right: '20px', top: '20px' },
        color: 'red',
        paddingRight: '30px',
        ':hover': {
          color: 'green',
        },
        paddingLeft: '30px',
      }),
    ).toBe('a b');

    expect(nonAtomic.getStyleMarkup()).toMatchSnapshot();
  });
  it('deletes properties', () => {
    const server = new GlitzServer<TestStyle>();

    expect(
      server.injectStyle({
        color: 'red',
        paddingRight: '10px',
        padding: { right: undefined },
        animationName: { from: { color: 'red' }, to: { color: 'green' } },
        animation: { name: undefined },
      }),
    ).toBe('a');

    expect(
      server.injectStyle([
        {
          color: 'red',
          paddingRight: '10px',
        },
        {
          paddingRight: (null as any) as undefined,
        },
      ]),
    ).toBe('a');

    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('applies transformer', () => {
    const server = new GlitzServer<TestStyle>({
      transformer: properties => {
        const prefixed: Properties = {};
        let property: keyof Properties;
        for (property in properties) {
          const value = properties[property];
          if (property === 'appearance' && value === 'none') {
            prefixed.MozAppearance = value;
          }
          prefixed[property] = properties[property] as Properties[typeof property];
        }
        return prefixed;
      },
    });

    expect(server.injectStyle({ appearance: 'none', animationName: { from: { appearance: 'none' } } })).toBe('a b');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
  it('passes theme', () => {
    const server = new GlitzServer<TestStyle>();

    expect(server.injectStyle({ color: (theme: any) => theme.text }, { text: 'red' })).toBe('a');
    expect(server.getStyleMarkup()).toMatchSnapshot();
  });
});
