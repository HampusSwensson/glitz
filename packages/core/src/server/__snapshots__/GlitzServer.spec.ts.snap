// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server applies transformer 1`] = `"<style class=\\"__glitz__\\">.a{-moz-appearance:none;appearance:none}.b{animation-name:a}@keyframes a{from{-moz-appearance:none;appearance:none}}</style>"`;

exports[`server deletes properties 1`] = `"<style class=\\"__glitz__\\">.a{color:red}</style>"`;

exports[`server injects atomic rules 1`] = `"<style class=\\"__glitz__\\">.a{color:red}.b{background-color:green}.c{border-color:blue}.d:hover{color:red}.e:hover{background-color:green}.f:hover{border-color:blue}</style><style class=\\"__glitz__\\" media=\\"(min-width: 768px)\\">.g{color:red}.h{background-color:green}.i{border-color:blue}</style>"`;

exports[`server injects different combinations 1`] = `"<style class=\\"__glitz__\\">.a{color:red}</style><style class=\\"__glitz__\\" media=\\"(min-width: 768px)\\">.b{color:green}</style><style class=\\"__glitz__\\" media=\\"(min-width: 992px)\\">.c{color:blue}</style>"`;

exports[`server injects font face rule 1`] = `"<style class=\\"__glitz__\\">.a{font-family:a}.b{font-family:b}.c{font-family:b,sans-serif}.d{font-family:sans-serif}@font-face {font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2');font-family:a}@font-face {font-style:normal;font-weight:400;src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2');font-family:b}</style>"`;

exports[`server injects keyframes rule 1`] = `"<style class=\\"__glitz__\\">.a{animation-name:a}.b{animation-name:b}.c{animation-name:some-thing}@keyframes a{from{color:red}to{color:green}}@keyframes b{from{color:blue}to{color:white}}</style>"`;

exports[`server injects media markup in certain order 1`] = `"<style class=\\"__glitz__\\">.d{color:red}</style><style class=\\"__glitz__\\" media=\\"(min-width: 100px)\\">.b{color:red}</style><style class=\\"__glitz__\\" media=\\"(min-width: 200px)\\">.c{color:red}</style><style class=\\"__glitz__\\" media=\\"(min-width: 300px)\\">.a{color:red}</style>"`;

exports[`server injects media rule 1`] = `"<style class=\\"__glitz__\\">.a{color:red}.b:hover{color:red}</style><style class=\\"__glitz__\\" media=\\"(min-width: 768px)\\">.c{color:red}.d:hover{color:red}</style>"`;

exports[`server injects nested pseudo rule 1`] = `"<style class=\\"__glitz__\\">.a:first-child:hover{color:red}</style>"`;

exports[`server injects non-atomic rules 1`] = `"<style class=\\"__glitz__\\">.a{color:red;background-color:green;border-color:blue}.b:hover{color:red;background-color:green;border-color:blue}</style><style class=\\"__glitz__\\" media=\\"(min-width: 768px)\\">.c{color:red;background-color:green;border-color:blue}</style>"`;

exports[`server injects plain rule 1`] = `"<style class=\\"__glitz__\\">.a{color:red}</style>"`;

exports[`server injects pseudo rule 1`] = `"<style class=\\"__glitz__\\">.a{color:red}.b:hover{color:red}</style>"`;

exports[`server injects rule deeply 1`] = `"<style class=\\"__glitz__\\">.a{color:red}.b{padding-left:20px}.e{animation-name:a}.f{animation-name:b}.g{font-family:a}.h{font-family:b}.c:hover{color:red}.d:first-child:hover{color:red}@keyframes a{from{color:green}to{color:blue}}@keyframes b{from{color:white}to{color:black}}@font-face {src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTjgPvNiA.woff2) format('woff2');font-family:a}@font-face {src:url(https://fonts.gstatic.com/s/paytoneone/v10/0nksC9P7MfYHj2oFtYm2ChTtgPs.woff2) format('woff2');font-family:b}</style><style class=\\"__glitz__\\" media=\\"(min-width: 768px)\\">.i{color:red}.j:hover{color:red}</style>"`;

exports[`server injects shorthand rule 1`] = `"<style class=\\"__glitz__\\">.a{padding-left:10px}.b{padding-right:10px}.c{padding-top:10px}.d{padding-bottom:10px}.e{grid-column-gap:10px}.f{margin-left:10px}.g{margin-right:10px}.h{margin-top:10px}.i{margin-bottom:10px}.j{padding-left:30px}.k{animation-name:a}.l{font-family:a}@keyframes a{from{padding-left:20px}}@font-face {font-weight:bold;font-family:a}</style>"`;

exports[`server preserves order 1`] = `"<style class=\\"__glitz__\\">.a{padding-top:20px}.b{color:red}.c{padding-right:30px}.e{padding-left:30px}.d:hover{color:green}</style>"`;

exports[`server preserves order 2`] = `"<style class=\\"__glitz__\\">.a{padding-top:20px;color:red;padding-right:30px;padding-left:30px}.b:hover{color:green}</style>"`;
